src/flamenco/runtime/program/fd_bpf_loader_v3_program.c

```diff
@@ -71,1471 +94,1399 @@ read_bpf_upgradeable_loader_state_for_program( fd_exec_txn_ctx_t * txn_ctx,
+    }
 
-    if (new_len > MAX_PERMITTED_DATA_LENGTH) {
-      return FD_EXECUTOR_INSTR_ERR_INVALID_REALLOC;
+    fd_bpf_upgradeable_loader_state_t programdata_state = {0};
+    fd_pubkey_t * upgrade_authority_address = NULL;
+    err = fd_bpf_loader_v3_program_read_state( instr_ctx, programdata_account, &programdata_state );
+    if( FD_UNLIKELY( err!=FD_BINCODE_SUCCESS ) ) {
+      FD_LOG_WARNING(( "Bpf state read for programdata account failed" ));
+      return err;
     }
+    if( fd_bpf_upgradeable_loader_state_is_program_data( &programdata_state ) ) {
+      if( FD_UNLIKELY( clock.slot==programdata_state.inner.program_data.slot ) ) {
+        FD_LOG_WARNING(( "Program was extended in this block already" ));
+        return FD_EXECUTOR_INSTR_ERR_ACC_IMMUTABLE;
+      }
 
-    fd_sol_sysvar_clock_t clock;
-    fd_sysvar_clock_read( &clock, ctx.slot_ctx );
-    ulong clock_slot = clock.slot;
+      if( FD_UNLIKELY( !programdata_state.inner.program_data.upgrade_authority_address ) ) {
+        FD_LOG_WARNING(( "Cannot extend ProgramData accounts that are not upgradeable" ));
+        return FD_EXECUTOR_INSTR_ERR_INVALID_ACC_DATA;
+      }
+      upgrade_authority_address = programdata_state.inner.program_data.upgrade_authority_address;
+    } else {
+      FD_LOG_WARNING(( "ProgramData state is invalid" ));
+      return FD_EXECUTOR_INSTR_ERR_INVALID_ACC_DATA;
+    }
+
+    /* https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L1232-L1256 */
+    ulong balance          = programdata_account->const_meta->info.lamports;
+    ulong min_balance      = fd_rent_exempt_minimum_balance( instr_ctx->slot_ctx, new_len );
+    min_balance            = fd_ulong_if( min_balance>0UL, min_balance, 1UL );
+    ulong required_payment = fd_ulong_sat_sub( min_balance, balance );
+
+    if( FD_UNLIKELY( required_payment>0UL ) ) {
+      fd_pubkey_t const * payer_key = &txn_accs[ instr_acc_idxs[ 3UL ] ];
+      
+      fd_system_program_instruction_t instr = {
+        .discriminant   = fd_system_program_instruction_enum_transfer,
+        .inner.transfer = required_payment
+      };
+
+      fd_vm_rust_account_meta_t * acct_metas = (fd_vm_rust_account_meta_t *)fd_scratch_alloc( 
+                                                                              FD_VM_RUST_ACCOUNT_META_ALIGN, 
+                                                                              2UL * sizeof(fd_vm_rust_account_meta_t) );
+      fd_native_cpi_create_account_meta( payer_key,       1UL, 1UL, &acct_metas[ 0UL ] );
+      fd_native_cpi_create_account_meta( programdata_key, 0UL, 1UL, &acct_metas[ 1UL ] );
+
+      fd_pubkey_t signers[ 1UL ];
+      signers[ 0UL ] = *payer_key;
+
+      err = fd_native_cpi_execute_system_program_instruction(
+        instr_ctx,
+        &instr,
+        acct_metas,
+        2UL,
+        signers,
+        1UL
+      );
+      if( FD_UNLIKELY( err ) ) {
+        return err;
+      }
+    }
```

Comment:
Agave doesn't pass any signers: https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L1250-L1254

Reply:
Looks like GitHub put this comment in the wrong place - I was referring to https://github.com/firedancer-io/firedancer/blob/00bf781b080b7a37ac82bf4b588048d31ba820bb/src/flamenco/runtime/program/fd_bpf_loader_v3_program.c#L1321-L1322